### PR TITLE
sync lifetime implementation (c# and kotlin parts) + tests

### PR DESCRIPTION
**Pull request by Iliya-usov**

### CODE REVIEW

File: rd-net/Lifetimes/Lifetimes/OuterLifetime.cs [hunk 1]
```diff
@@ -113,7 +114,7 @@ public static LifetimeDefinition Define(OuterLifetime lifetime, string? id = nul
 
         // Attach as nested to the parent lifetime
 //        if(!nested.IsTerminated) // Might have been terminated by FAtomic
-        lifetime.Def.Attach(nested); // Pass True: might be terminated async on another thread between our check and AttachNested body (example: Queue from another thread)
+        lifetime.Def.Attach(nested, true); // Pass True: might be terminated async on another thread between our check and AttachNested body (example: Queue from another thread)
```
Comment (ForNeVeR) [inline rd-net/Lifetimes/Lifetimes/OuterLifetime.cs:117]: _(Minor)_ Any idea what this "Pass True" comment may be referring to?

After this change, it looks like it's talking about the second parameter of `Attach`, which, seemingly, cannot be?

File: rd-net/Lifetimes/Lifetimes/OuterLifetime.cs [hunk 2]
```diff
@@ -127,13 +128,24 @@ public static LifetimeDefinition Define(OuterLifetime lifetime, string? id = nul
         
     /// <summary>
     /// Creates an intersection of some lifetimes — a lifetime to terminate when either one terminates.
+    /// /// Created lifetime inherits the smallest <see cref="Lifetime.TerminationTimeoutKind"/>
```
Comment (ForNeVeR) [inline rd-net/Lifetimes/Lifetimes/OuterLifetime.cs:131]: _(Minor)_
```suggestion
    /// Created lifetime inherits the smallest <see cref="Lifetime.TerminationTimeoutKind"/>
```

**Review by ForNeVeR — CHANGES_REQUESTED** [review]

File: rd-kt/rd-core/src/main/kotlin/com/jetbrains/rd/util/lifetime/RLifetime.kt [hunk 3]
```diff
@@ -261,16 +383,23 @@ class LifetimeDefinition : Lifetime() {
...
 
         //wait for all executions finished
-        if (!spinUntil(waitForExecutingInTerminationTimeout) { executingSlice[state] <= threadLocalExecuting[this] }) {
-            log.error { "$this: Can't wait for executeIfAlive for more than $waitForExecutingInTerminationTimeout ms. Keep termination." }
+        val terminationTimeoutMs = getTerminationTimeoutMs(terminationTimeoutKind)
+        if (!spinUntil(terminationTimeoutMs) { executingSlice[state] <= threadLocalExecuting[this] }) {
+            log.warn {
+                "$this: can't wait for `executeIfAlive` completed on other thread in $terminationTimeoutMs ms. Keep termination.${System.lineSeparator()}" +
+                "This may happen either because of the executeIfAlive failed to complete in a timely manner. In the case there will be following error messages.${System.lineSeparator()}" +
+                "Or this might happen because of garbage collection or when the thread yielded execution in SpinWait.SpinOnce but did not receive execution back in a timely manner. If you are on JetBrains' Slack see the discussion https://jetbrains.slack.com/archives/CAZEUK2R0/p1606236742208100"
```
[the remark this answers is not in the document]
Comment (Iliya-usov) [reply rd-kt/rd-core/src/main/kotlin/com/jetbrains/rd/util/lifetime/RLifetime.kt:394]: I added an additional condition check to spinUntil and now this message is not needed

File: rd-net/Lifetimes/Lifetimes/OuterLifetime.cs (hunk 1 above)
Comment (Iliya-usov) [reply rd-net/Lifetimes/Lifetimes/OuterLifetime.cs:117]: I think this comment refers to the commented ` if(!nested.IsTerminated)`. I think we can just delete these commented lines

**Review by ForNeVeR — APPROVED** [review]
🐐

**Pull request merged**